File: api/Models/HubMessage.cs
using System.Text.Json.Serialization;

namespace CroApp.Api.Models;

// One row per bird that has ever landed at a Hub - written once, at arrival, by
// BirdService.ResolveArrivalIfDueAsync, and never updated afterward. Deliberately durable
// independent of the underlying Bird's own lifecycle: unlike GetHubResidentsAsync (a live
// "who's currently sitting here" snapshot), a HubMessage survives the bird later being
// picked up and resent elsewhere, or even deleted. HubId is the partition key - the
// dominant query is "list every message posted to a given Hub, newest first", same
// reasoning as Hub's own /status and BirdReaction's /birdId. The container's
// DefaultTimeToLive (see Program.cs) auto-expires rows 7 days after CreatedAt, so the
// board resets itself with no cleanup job. SenderProfilePictureUrl is deliberately NOT
// stored here - resolved live at read time via the same N+1 pattern GET /friends already
// uses, so an avatar changed after posting still shows up-to-date on old board entries.
public record HubMessage(
    [property: JsonPropertyName("id")] string Id,
    string HubId,
    string BirdId,
    string SenderId,
    string SenderUsername,
    string BirdName,
    string? OriginNestName,
    string Type,
    string? Content,
    string? AudioUrl,
    string? ImageUrl,
    DateTimeOffset CreatedAt);

File: app/lib/models/hub_message.dart
// One entry on a Hub's public message board - a durable record of a bird that landed
// there, independent of that bird's own later lifecycle (it may since have been picked up
// and resent elsewhere). See api/Models/HubMessage.cs.
class HubMessage {
  final String id;
  final String senderId;
  final String senderUsername;
  final String? senderProfilePictureUrl;
  final String birdName;
  final String? originNestName;
  final String type;
  final String? content;
  final String? audioUrl;
  final String? imageUrl;
  final DateTime createdAt;

  HubMessage({
    required this.id,
    required this.senderId,
    required this.senderUsername,
    this.senderProfilePictureUrl,
    required this.birdName,
    this.originNestName,
    required this.type,
    this.content,
    this.audioUrl,
    this.imageUrl,
    required this.createdAt,
  });

  factory HubMessage.fromJson(Map<String, dynamic> json) => HubMessage(
        id: json['id'] as String,
        senderId: json['senderId'] as String,
        senderUsername: json['senderUsername'] as String,
        senderProfilePictureUrl: json['senderProfilePictureUrl'] as String?,
        birdName: json['birdName'] as String,
        originNestName: json['originNestName'] as String?,
        type: json['type'] as String,
        content: json['content'] as String?,
        audioUrl: json['audioUrl'] as String?,
        imageUrl: json['imageUrl'] as String?,
        createdAt: DateTime.parse(json['createdAt'] as String),
      );
}
